### ISSUE TITLE
Docker hangs indefinitely when issuing a docker wait for non-existent container

### ISSUE THREAD

**Issue by cblecker** [moby/moby#33948]
**Description**
After updating to Docker 17.06.0-ce, there is a functionality change in the `docker wait` command compared to 17.05, when issued against a container name that doesn't exist. I can't seem to find any documentation on this change, and am unsure if it's a bug (it appears to be).

**Steps to reproduce the issue:**
1. Upgrade to Docker 17.06.0-ce
2. Issue `docker wait` for a container that doesn't exist
3. Docker will hang.

**Describe the results you received:**
In 17.06, if you issue a `docker wait` for a container name that doesn't exist, the docker command will hang indefinitely.

**Describe the results you expected:**
In 17.05, if you issue a `docker wait` for a container name that doesn't exist, you get a non-zero exit with the following error:
```
$ docker wait kube-rsync-5095346036-5-v1.8.3-2
Error response from daemon: No such container: kube-rsync-5095346036-5-v1.8.3-2
```

Similarly, if you issue a `docker kill` in either 17.06 or 17.05, you get a similar error (what I expect to be desired functionality):
```
docker kill kube-rsync-5095346036-5-v1.8.3-2
Error response from daemon: Cannot kill container kube-rsync-5095346036-5-v1.8.3-2: No such container: kube-rsync-5095346036-5-v1.8.3-2
```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:31:53 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:51:55 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 17.06.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.31-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952GiB
Name: moby
ID: ZV2V:GX2D:5RZF:47YT:NFFB:TPFP:6KBC:QFF3:QJSN:EUAK:PKOP:3NVP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 17
 Goroutines: 28
 System Time: 2017-07-04T20:07:55.056942415Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by thaJeztah** [comment]
I'm able to reproduce this.

Adding a debug-line to see where it fails;

```patch
diff --git a/api/server/router/container/container_routes.go b/api/server/router/container/container_routes.go
index 96b1010e1..3e3ae9175 100644
--- a/api/server/router/container/container_routes.go
+++ b/api/server/router/container/container_routes.go
@@ -308,6 +308,7 @@ func (s *containerRouter) postContainersWait(ctx context.Context, w http.Respons
        // router.WithCancel() wrapper.
        waitC, err := s.backend.ContainerWait(ctx, vars["name"], waitCondition)
        if err != nil {
+               logrus.Errorf("Handler for %s %s returned error: %v", r.Method, r.URL.Path, err)
                return err
        }
```

It looks like it's actually returning that error;

```
DEBU[0005] Calling GET /_ping
DEBU[0005] Calling POST /v1.30/containers/foo/wait?condition=
ERRO[0005] Handler for POST /v1.30/containers/foo/wait returned error: No such container: foo
```


Both API 1.30 and 1.29 arrive at the same error;

```
DEBU[0005] Calling GET /_ping
DEBU[0005] Calling POST /v1.30/containers/foo/wait?condition=
ERRO[0005] Handler for POST /v1.30/containers/foo/wait returned error: No such container: foo
DEBU[0078] Calling GET /_ping
DEBU[0078] Calling POST /v1.29/containers/foo/wait
ERRO[0078] Handler for POST /v1.29/containers/foo/wait returned error: No such container: foo
```


As a temporary workaround, perhaps it's possible to  override the API version that's being used; using API version v1.29 does handle the error;

```bash
DOCKER_API_VERSION=v1.29 docker wait foo
Error response from daemon: {"message":"No such container: foo"}
```

(apart from printing the literal JSON body, instead of just the message 😅 )


ping @jlhawn any ideas?

**Comment by thaJeztah** [comment]
This was actually already fixed in https://github.com/moby/moby/pull/33293, which is in docker/cli master, but not yet in docker 17.06.0. I created a tracking issue to include the fix into 17.06.1

**Comment by cblecker** [comment]
Thank you for looking into this, @thaJeztah!

**Comment by cblecker** [comment]
Fixed in 17.06.1 by https://github.com/docker/docker-ce/pull/106
Thanks @thaJeztah !
